onnxruntime/core/providers/webgpu/shader_variable.h

```diff
@@ -0,0 +1,338 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <sstream>
+#include <set>
+
+#include "core/framework/tensor_shape.h"
+
+#include "core/providers/webgpu/program.h"
+
+namespace onnxruntime {
+namespace webgpu {
+
+template <typename TIdx,
+          typename TRank,
+          typename = std::enable_if_t<std::is_same_v<TRank, int> || std::is_same_v<TRank, size_t>>>
+std::string GetElementAt(std::string_view var, const TIdx& idx, TRank rank, bool is_f16 = false) {
+  // "std::string::rfind(str, 0) == 0" is equivalent to "std::string::starts_with(str)" before C++20.
+  if (var.rfind("uniforms.", 0) == 0) {
+    if (rank > 4) {
+      if constexpr (std::is_integral_v<TIdx>) {
+        if (is_f16) {
+          return MakeStringWithClassicLocale(var, "[", idx / 8, "][", (idx % 8) / 4, "][", (idx % 8) % 4, "]");
+        } else {
+          return MakeStringWithClassicLocale(var, "[", idx / 4, "][", idx % 4, "]");
+        }
+      } else {
+        if (is_f16) {
+          return MakeStringWithClassicLocale(var, "[(", idx, ") / 8][(", idx, ") % 8 / 4][(", idx, ") % 8 % 4]");
+        } else {
+          return MakeStringWithClassicLocale(var, "[(", idx, ") / 4][(", idx, ") % 4]");
+        }
+      }
+    }
+  }
+
+  return rank > 1 ? MakeStringWithClassicLocale(var, "[", idx, "]") : std::string{var};
+}
+
+struct ShaderUsage {
+  enum : uint32_t {
+    None = 0,                             // no usage. this means no additional implementation code will be generated.
+    UseIndicesTypeAlias = 1,              // use type alias "{name}_indices_t" for indices (eg. u32, vec2<u32>, vec3<u32>, vec4<u32>, ...)
+    UseValueTypeAlias = 2,                // use type alias "{name}_value_t" for value (eg. f32, vecT<f32>, vec4<bool>, ...)
+    UseElementTypeAlias = 4,              // use type alias "{name}_element_t" for element (eg. f32, bool, ...)
+    UseShapeAndStride = 16,               // use shape and stride for the variable
+    UseOffsetToIndices = 32,              // use implementation of fn o2i_{name}
+    UseIndicesToOffset = 64,              // use implementation of fn i2o_{name}
+    UseBroadcastedIndicesToOffset = 128,  // use implementation of fn {broadcasted_result_name}_bi2o_{name}
+    UseSet = 256,                         // use implementation of fn set_{name}
+    UseSetByIndices = 512,                // use implementation of fn set_{name}_by_indices
+    UseGet = 1024,                        // use implementation of fn get_{name}
+    UseGetByIndices = 2048,               // use implementation of fn get_{name}_by_indices
+    UseUniform = 32768,                   // use uniform for shape and stride
+  } usage;
```

Comment:
Actually here the member var `usage` is not an enum. You use it as an integer. Its value is not limited to the ones that are listed here. Its value may be any combination of them.

Reply:
I think performance is not super critical here. I would prefer using std::bitset<16> instead. Then you do not need to worry about shifting and bitmask operations, which can make the code easier to read.

Reply:
Yes it is used as flag enum. is there any best practices that can be applied to the code here?

Reply:
```
enum Flags{
  flag1,
  flag2,
  flag3,
  ...
}


std::bitset<16> usage;
uages.set(flag1);
uages.set(flag3);
uages.test(flag1)
int value = uages.to_ulong();
//Then pass the value to dawn API.

I personally feel this is more readable and safer, but it is up to you to decide which is better.
 

Reply:
how do you suggest to design the interface of this:

```
AddInput( inputTensor, flag1 | flag2 | flag3 );
```

Reply:
```
using ShaderUsage  = std::bitset<16>;
```
The second parameter of the two functions is a bitset. 
But, then we will not able to do 
```
const auto& x = shader.AddInput("x", ShaderUsage::UseUniform | ShaderUsage::UseValueTypeAlias);
```
in one line. 
Fine. Let's discuss it later.